### PR TITLE
Add async storage for saving/getting large data [DO NOT MERGE]

### LINKED ISSUE #3239: Switch to async storage (i.e. indexdb instead of localstorage)
Currently synchronous localstorage is used, which is fine for most things apart from [saving history](https://github.com/openstreetmap/iD/blob/39c82af15891bc29f89502ed60a3498895a1353e/modules/core/history.js#L372) which can grow very large and causes significant slow-down, despite the being [debounced to 350ms](https://github.com/openstreetmap/iD/blob/f7e49e79f8a6e5e2be33bca17ddfb3b1f4a39c11/js/id/id.js#L349-L356).

This is most noticeable when editing/creating complex line/polygon features, which result in the history being large, and writes to localstorage taking significant time, locking up the browser every 350ms. iD quickly becomes unusable in this scenario. This can be solved by frequent committing, but it is not an ideal solution.

To avoid a larger re-factor which would necessitate making all storage calls async, perhaps the quickest/easiest solution would be to use async storage for only `history.save()`.

IndexedDB is the obvious choice. Any recommendations for a lightweight wrapper that gets around any IndexedDB bugs or lack of support on browsers currently supported by iD (looking at you, Safari)? We only need a wrapper that implements basic `get()` and `put()`.

### PR DESCRIPTION
**Pull request by Semigradsky**
Fix #3239

### CODE REVIEW

File: modules/core/history.js [hunk 1]
```diff
@@ -577,18 +577,27 @@ export function coreHistory(context) {
...
 
         // is iD not open in another window and it detects that
-        // there's a history stored in localStorage that's recoverable?
+        // there's a history stored in async storage that's recoverable?
         restorableChanges: function() {
-            return lock.locked() && !!context.storage(getKey('saved_history'));
+            var isLocked = lock.locked();
+
+            if (isLocked) {
+                return Promise.resolve(true);
```
Comment (bhousel) [inline modules/core/history.js]: I'm pretty sure native Promises don't work in IE11.. Can you use localForage's callback API instead?

Comment (Semigradsky) [reply modules/core/history.js]: Oh, my fault. Rewrited to callback API.

**Comment by Semigradsky** [comment]
@bhousel ping

File: modules/core/context.js [hunk 2]
```diff
@@ -431,7 +455,7 @@ export function coreContext() {
 
     // Debounce save, since it's a synchronous localStorage write,
     // and history changes can happen frequently (e.g. when dragging).
-    context.debouncedSave = _debounce(context.save, 350);
+    context.debouncedSave = _debounce(context.save, 100);
```
Comment (bhousel) [inline modules/core/context.js]: Is there a reason why you changed this?  (If anything I think we could slow down the saves, not speed them up)

**Comment by bhousel** [comment]
Thanks for being patient @Semigradsky - I think I'm going to move this new code into its own module, and also try to update the other places in iD that use storage to use this new store, and add some code to grab the existing localStorage values if they exist.  Hope to merge it today 👍

**Comment by bhousel** [comment]
Hey @Semigradsky - I did some work on this yesterday but here's where I ended up:

- The switch from sync to async storage is a pretty big deal.  We can do it, (and there are real benefits) but it changes the flow of when stored variables can be available to the iD code
- IndexedDB is actually kind of slow.  I'm reading that get times around like 100-300ms are normal for largish objects.
- Some components in iD need access to the stored variables at startup (think like the splash screen and the history restore)  - however we don't have an ability to _delay_ startup until the stored variables are ready - this would be an API breaking change..
- So considering the current code and that it would be a breaking change, I want to put the pause on this PR until we start work on iD v3.

I don't have a timeline on when it would happen, but it is something we want eventually..

**Pull request closed without merging**